discovery/hyperspace/chunk.go

```diff
@@ -0,0 +1,14 @@
+package hyperspace
+
+func chunk(b []byte, n int) [][]byte {
```

Comment:
can you explain the logic of this?

Reply:
The idea behind this whole thing is to create a sparse vector with representations of all some keys a peer wants to advertise.

So, let's say we want to end up with a vector with 300 dimensions. (Chosen with fair roll of the dice)

* Take a key (of let's say arbitrary length)
* Split it in chunks (of let's say 2bytes each)
* Convert each chunk into an int (we currently use murmur for that which is an overkill)
* — In the current implementation, each chunk is prefixed with the “type” of the value being chunked, eg peer key, identity key, protocol, content hash, etc. I'm not sure this is reasonable, and probably we should require all keys to be hashes and be done with the prefix.
* Assuming that the each numeric chunk cannot fit in the low number of dimensions, we normalise it between 0 and 300.
* We now have 16 numbers between 0 and 300 that represent our hash. This is our sparse vector. We'll call this a key vector for now.

We can now merge all our key vectors into a composite one. Due to the low dimensionality of our vectors, many key vectors will be occupying the same dimensions, but that's fine as it allows reducing the space. Changing the number of dimensions and the size of each chunk allows control collisions. We need to make sure that we don't have too few dimensions so that the vector is so "full" that large parts of the hash vectors collide. How we decide on the vector size and how we can figure out how many keys a vector can hold before it becomes saturated and no longer useful; I honestly have no clue.

cc @dimmu